src/FluffySpoon.AspNet.LetsEncrypt.Aws/AwsDnsChallengePersistenceStrategy.cs

```diff
@@ -0,0 +1,204 @@
+		{
+			_logger.LogDebug("Finding hosted zone responsible for {DnsName}", dnsName);
+
+			var partsToMatch = dnsName.Split(DomainSegmentSeparator);
+			var isWildcard = dnsName.StartsWith(WildcardPrefix);
+			var bestPossibleScore = isWildcard ? partsToMatch.Length - 1 : partsToMatch.Length;
+
+			HostedZone bestMatch = null;
+			int bestMatchScore = 0;
+
+			// Enumerate all hosted zones until a match is found, or until all hosted zones have been enumerated
+			var hostedZones = await _route53Client.ListHostedZonesAsync();
+			while (hostedZones.IsTruncated)
+			{
+				foreach (var hostedZone in hostedZones.HostedZones)
+				{
+					_logger.LogDebug("Checking zone {Zone}", hostedZone.Name);
+
+					if (dnsName.ToLower().EndsWith(DomainSegmentSeparator + hostedZone.Name.ToLower()))
+					{
+						// Matches, calculate score
+						var hostedZoneParts = hostedZone.Name.Split(DomainSegmentSeparator);
+						var score = hostedZoneParts.Length;
+
+						if (score == bestPossibleScore)
+						{
+							_logger.LogInformation("Exact match for {DnsName} found (zone {Zone})", dnsName, hostedZone.Name);
+
+							// Exact match found
+							return hostedZone;
+						}
+						else if (score > bestMatchScore)
+						{
+							_logger.LogDebug("Setting best match for {DnsName} to zone {Zone}", dnsName, hostedZone.Name);
+
+							bestMatch = hostedZone;
+							bestMatchScore = score;
+						}
+					}
+				}
+
+				hostedZones = await _route53Client.ListHostedZonesAsync(new ListHostedZonesRequest() { Marker = hostedZones.Marker });
+			}
+
+			if (bestMatch == null)
+				_logger.LogInformation("No zone match for {DnsName} found", dnsName);
+			else
+				_logger.LogInformation("Best match for {DnsName} found (zone {Zone})", dnsName, bestMatch.Name);
+
+			return bestMatch;
+		}
+
+		private async Task<IEnumerable<ResourceRecordSet>> FindRecordSets(HostedZone zone, string dnsName, string recordType)
+		{
+			_logger.LogDebug("Finding record sets for {RecordType} {DnsName} in zone {Zone}", recordType, dnsName, zone.Name);
+
+			var result = new List<ResourceRecordSet>();
+			var rootedDnsName = dnsName.EndsWith(DomainSegmentSeparator.ToString()) ? dnsName : dnsName + DomainSegmentSeparator;
+			var remainder = dnsName.Replace(zone.Name, String.Empty);
+			var recordSets = await _route53Client.ListResourceRecordSetsAsync(new ListResourceRecordSetsRequest() { HostedZoneId = zone.Id, StartRecordType = RRType.FindValue(recordType), StartRecordName = dnsName });
+
+			while (recordSets.IsTruncated)
+			{
+				foreach (var recordSet in recordSets.ResourceRecordSets)
+				{
+					if (recordSet.Name.ToLower().Equals(rootedDnsName.ToLower()))
+						result.Add(recordSet);
+				}
+
+				recordSets = await _route53Client.ListResourceRecordSetsAsync(new ListResourceRecordSetsRequest() { HostedZoneId = zone.Id, StartRecordType = recordSets.NextRecordType, StartRecordName = recordSets.NextRecordName, StartRecordIdentifier = recordSets.NextRecordIdentifier });
+			}
```

Comment:
As before, it would just result in a method of the same size, just a different name...

Reply:
As above, I've reviewed it a few times, and because of the interaction between variables, I've not come up with a way to simplify/refactor it in any constructive way.